Purchases: Use `state.purchases` to render `/purchases`

drewblaisdell opened the pull request:
This is part of a series of PRs to replace `PurchasesStore` with a new branch of global state.

This PR updates the following components to use actions and data from `state/purchases/` instead of `PurchasesStore` and Flux actions:

- `PurchasesList`
- `ManagePurchase`
- `CancelPurchase`
- `CancelPrivateRegistration`
- `ConfirmCancelDomain`
- `RemovePurchase`
- `EditCardDetails`

It also removes the now unused actions and data components that were used in these components, and replaces an instance of `clearPurchases` in `Checkout`.

**Note:** This is large. ~80 lines are commits that will be handled separately, and will be removed before merging. I tried to make the revision history easy to read, and would recommend looking at the commits individually.

**Testing**
`PurchasesList`
- Visit `/purchases`.
- Assert that you see placeholders for a list of purchases.
- Assert these placeholders are eventually replaced with your actual purchases.

`ManagePurchase`
- Click on a purchase on `/purchases`.
- Assert that you are taken to a page with information about that purchase.
- Refresh the page.
- Assert that you see a loading placeholder.
- Assert that the placeholder is eventually replaced with information about the purchase.

`CancelPurchase`
- Visit `/purchases` and click on a renewable purchase that is outside the refund period.
- Click the `Cancel Subscription` link.
- Click the `Cancel Subscription` button.
- Assert that you are redirected to `/purchases` with a success notice and that the purchases are fetching again.

`CancelPrivateRegistration`
- Apply D2256-code and ensure that your hosts file points to your sandbox.
- Visit `/purchases` and click on a domain purchase with a private registration.
- Click the `Cancel Private Registration` link.
- Click the `Cancel Private Registration` button.
- Assert that you see an error message that reads `foo Please try again later or contact support.` (`foo` is just a generic string the API returns in my patch as an error message).
- Revert the patch.
- Click the `Cancel Private Registration` button again.
- Assert that you are redirected to `ManagePurchase` for this domain with a success notice.

`ConfirmCancelDomain`
- Visit `/purchases` and click on a domain purchase that is within the refund period.
- Click the `Cancel Domain and Refund` link.
- Click the `Cancel Domain and Refund` button.
- Select a reason to cancel and click `Cancel Anyway`.
- Assert that you are redirected to `/purchases` with a success notice and that the purchases are fetching again.

`RemovePurchase`
- Visit `/purchases` and click on a purchase that has auto-renew disabled.
- Click `Remove [purchase name]`.
- Fill out the survey and proceed through the dialog.
- Assert that you are redirected to the list of purchases, and that this purchase is not in that list.

`EditCardDetails`
- Visit `/purchases` and click on a purchase made with a credit card.
- Click `Edit Payment Method`.
- Assert that you see a placeholder as the list of the user's stored cards fetch.
- Assert that the page eventually loads.
- Enter and submit new credit card details.
- Assert that you are redirected to `ManagePurchase` for the given purchase, and that you see placeholders while the purchase is fetched again.

`Checkout`
- Visit `/purchases` and wait until the purchases have loaded.
- Purchase a plan in the same browser session (by clicking through `My Sites`, `Plan`, etc.).
- Once you reach the "thank you" page in checkout, navigate to `/purchases` in the same session.
- Assert that you see placeholders as the purchases load again, and that the purchase you just made appears in the list of purchases.

- [x] Code
- [x] Product

Tug commented on client/lib/upgrades/actions/purchases.js:
```diff
@@ -9,66 +9,24 @@ import i18n from 'i18n-calypso';
 
-const PURCHASES_FETCH_ERROR_MESSAGE = i18n.translate( 'There was an error retrieving purchases.' ),
-	PURCHASE_REMOVE_ERROR_MESSAGE = i18n.translate( 'There was an error removing the purchase.' );
+const PURCHASES_FETCH_ERROR_MESSAGE = i18n.translate( 'There was an error retrieving purchases.' );
```
Let's remove this definition

spncrb commented:
As far as I can tell this Product is working.  👍 


The few differences from testing instructions were as follows:
* During CancelPrivateRegistration When using the patch and cancelling private registration, I got the normal messaging, landed on manage purchase, but the privacy was never disabled: https://cloudup.com/c7teaXcNazq
* During RemovePurchase did not leave me with a blank list, showed my purchases still https://cloudup.com/c3fyRtnSWCS
* During EditCardDetails steps, Edit payment results in ONLY a new card form, no way to select already saved cards: https://cloudup.com/czy3qBYDQTL

I think these are unrelated or intentional, but noting them here if not.

Tug commented on client/lib/upgrades/actions/purchases.js:
```diff
@@ -9,66 +9,21 @@ import i18n from 'i18n-calypso';
-const PURCHASES_FETCH_ERROR_MESSAGE = i18n.translate( 'There was an error retrieving purchases.' ),
-	PURCHASE_REMOVE_ERROR_MESSAGE = i18n.translate( 'There was an error removing the purchase.' );
-
 function cancelPurchase( purchaseId, onComplete ) {
```
We can move `cancelPurchase` as well, can't we?
I think only `me/purchases/cancel-purchase` would need to be updated which is in the scope of this PR.

drewblaisdell replied:
Do you mean create a Redux action for it? I agree that we should do this, but it probably makes sense to do it separately, as this PR just intends to remove the old Flux actions/store and hook up the new Redux actions/branch of state.

Tug replied:
Well yes, but `cancelPurchase` used to call `clearPurchases` which was dispatching `PURCHASES_REMOVE`. Now that it doesn't its behavior has changed I believe. There is only `cancelPurchase` and `cancelAndRefundPurchase` left to move so we have everything from `purchases` (which was the original goal of thos PR I believe).

drewblaisdell replied:
> Now that it doesn't its behavior has changed.

:+1: I updated the components that call `cancelPurchase` to call `clearPurchases` in the `cancelPurchase` callback to ensure that the behavior of this is the same in the meantime.

stephanethomas commented on client/components/data/purchases/edit-card-details/index.jsx:
```diff
@@ -7,47 +7,34 @@ import React from 'react';
+import { fetchStoredCards } from 'lib/upgrades/actions';
 import observe from 'lib/mixins/data-observe';
-import PurchasesStore from 'lib/purchases/store';
-import { shouldFetchUserPurchases } from 'lib/purchases';
```
It looks like we can also retire this `shouldFetchUserPurchases()` function now.

Tug commented on client/lib/upgrades/actions/purchases.js:
```diff
@@ -118,72 +73,13 @@ function fetchStoredCards() {
-	} );
-}
-
 function cancelAndRefundPurchase( purchaseId, data, onComplete ) {
```
Same as https://github.com/Automattic/wp-calypso/pull/6647/files#r70770759

Tug commented on client/me/purchases/utils.js:
```diff
@@ -9,8 +9,9 @@ import page from 'page';
 
+// TODO: Remove these property-masking functions in favor of selectors
 function getPurchase( props ) {
-	return props.selectedPurchase.data;
```
This `utils` file doesn't work on the internal structure of the purchase anymore, it only works on the props. i think we could remove this file and access the props directly as it just make the code harder to follow.

drewblaisdell replied:
Yeah, I added the `TODO` on the right for this, although you're right, we should just get rid of these altogether instead of using selectors. I think we should do it separately, as this PR is large.

Tug commented:
I think it still needs a bit of cleaning/refactoring (see comments), otherwise it looks good to me

stephanethomas commented on client/me/purchases/controller.jsx:
```diff
@@ -201,15 +216,15 @@ export default {
+				purchaseId={ parseInt( context.params.purchaseId, 10 ) }
 				destinationType={ context.params.destinationType }
-				sites={ sites } />
+				user={ user.get() }
```
Do we really need to pass `user` here? It looks like the `ManagePurchase` component already retrieves such object on its own.

drewblaisdell replied:
Nice catch, I'll remove this from the props/propTypes. I'm not sure what I was thinking. :)

fabianapsimoes commented:
> During CancelPrivateRegistration When using the patch and cancelling private registration, I got the normal messaging, landed on manage purchase, but the privacy was never disabled: https://cloudup.com/c7teaXcNazq

I tested this, and it seems privacy actually _is_ disabled, but it takes a while for Purchases to reflect that.

> During RemovePurchase did not leave me with a blank list, showed my purchases still https://cloudup.com/c3fyRtnSWCS

@drewblaisdell I suppose the instructions are incorrect there? If the user had more than one upgrade, the list shouldn't be empty (though it should not show the removed purchase).

> During EditCardDetails steps, Edit payment results in ONLY a new card form, no way to select already saved cards: https://cloudup.com/czy3qBYDQTL

That's the expected behavior. Edit Payment Method only allow users to enter a new card, but not to manage / use stored ones.

drewblaisdell commented:
Thanks for reviewing, @spncrb, @Tug, @fabianapsimoes, @stephanethomas! I'll update this per your notes.

> @drewblaisdell I suppose the instructions are incorrect there? If the user had more than one upgrade, the list shouldn't be empty (though it should not show the removed purchase).

It shouldn't, but there is a bug where the list is cleared before being fetched again. With a small list of purchases on production, the `/me/purchases` response might be fast enough that it is hard to notice. I'll assert again when I update this that it is still happening.

drewblaisdell commented:
I rebased this, which revealed some issues introduced by #6706 because of some unexpected types from the API. I opened #6798 to address these, and cherry-picked those commits into this PR to make testing possible. Those commits will be removed per the status of that PR.

I also added [a commit](https://github.com/Automattic/wp-calypso/pull/6647/commits/e5f86122a9e9b5fc02d0950e7631eeacd8831243) to remove `PURCHASE_REMOVE`. @stephanethomas pointed out that we remove all purchases when this action is dispatched, which is erroneous. I updated the testing instructions to reflect this change.

There are a couple other refactors that have been mentioned that I think should be handled separately:

- Remove the property-masking functions in `client/me/purchases/utils`. There was some debate over whether these were ever helpful, but now that we have selectors, I think we all agree that they are redundant. :)
- Add actions for `cancelPurchase` and `cancelAndRefundPurchase`.

I'll open issues for these once this is merged.

Tug commented:
Merged #6798 and removed the 2 commits. Ok for doing those 2 remaining tasks later.
Code 👍

Tug commented:
Tests are failing, I need to fix that first

drewblaisdell commented:
I opened issues for the remaining tasks: https://github.com/Automattic/wp-calypso/issues/6821, https://github.com/Automattic/wp-calypso/issues/6822
